internal/blockchain/difficulty.go

```diff
@@ -212,20 +211,152 @@ func (b *BlockChain) calcNextRequiredDifficulty(prevNode *blockNode, newBlockTim
+	// relative calculations.
+	//
+	// However, since the initial difficulty is reset with the agenda, no
+	// additional offsets are needed.
+	timeDelta := prevNode.timestamp - blake3Anchor.timestamp
+	heightDelta := prevNode.height - blake3Anchor.height
+
+	// Calculate the next target difficulty using the ASERT algorithm.
+	//
+	// Note that the difficulty of the anchor block is NOT used for the initial
+	// difficulty because the difficulty must be reset due to the change to
+	// blake3 for proof of work.  The initial difficulty comes from the chain
+	// parameters instead.
+	params := b.chainParams
+	nextDiff := standalone.CalcASERTDiff(params.WorkDiffV2Blake3StartBits,
+		params.PowLimit, int64(params.TargetTimePerBlock.Seconds()), timeDelta,
+		heightDelta, params.WorkDiffV2HalfLifeSecs)
+
+	// Prevent the difficulty from going higher than a maximum allowed
+	// difficulty on the test network.  This is to prevent runaway difficulty on
+	// testnet by ASICs and GPUs since it's not reasonable to require
+	// high-powered hardware to keep the test network running smoothly.
+	//
+	// Smaller numbers result in a higher difficulty, so imposing a maximum
+	// difficulty equates to limiting the minimum target value.
+	if b.minTestNetTarget != nil && nextDiff < b.minTestNetDiffBits {
+		nextDiff = b.minTestNetDiffBits
+	}
+
+	return nextDiff
+}
+
+// blake3WorkDiffAnchor returns the block to treat as the anchor block for the
+// purposes of determining how far ahead or behind the ideal schedule the
+// provided block is when calculating the blake3 target difficulty for the block
+// AFTER the passed previous block node.
+//
+// This function MUST only be called with the blake3 proof of work agenda active
+// and with the chain state lock held (for writes).
+func (b *BlockChain) blake3WorkDiffAnchor(prevNode *blockNode) *blockNode {
+	// Use the previously cached anchor when it exists and is actually an
+	// ancestor of the passed node.
+	//
+	// Note that it will always be set to the genesis block on networks where
+	// the agenda is always active.
+	anchor := b.cachedBlake3WorkDiffAnchor.Load()
+	if anchor != nil && anchor.IsAncestorOf(prevNode) {
+		return anchor
+	}
+
+	// Find the block just prior to the activation of the blake3 proof of work
+	// agenda from the perspective of the block AFTER the passed block.
+	//
+	// The state of an agenda can only change at a rule change activation
+	// boundary, so determine the final block of the rule change activation
+	// interval just prior to the interval the block AFTER the provided one
+	// falls in and then loop backwards one interval at a time until the agenda
+	// is no longer active.
+	rcai := int64(b.chainParams.RuleChangeActivationInterval)
+	svh := b.chainParams.StakeValidationHeight
+	finalNodeHeight := calcWantHeight(svh, rcai, prevNode.height+1)
+	candidate := prevNode.Ancestor(finalNodeHeight)
+	for candidate != nil && candidate.parent != nil {
```

Comment:
Correct.  Similar to the above discussion, the agenda is either always active in which case it is enforced starting from block 1 or it must go through the activation process which requires stake validation height + a voting interval + a locked in interval.

Reply:
After further consideration, I don't think using the genesis block as the anchor for networks where the agenda is always active is the best choice.

I say this because the timestamp in the genesis has to be set to something when the code creates it, but by the time the code is merged and released, a long period of time might have passed.  That means that blocks based on it as the anchor will heavily be behind schedule leading to a major difficulty drop for a ton of initial blocks.

That doesn't matter too much for the simulation network where blocks being the easiest difficulty is the norm anyway, but it could very well be an issue for future versions of the test network.

Instead, I think networks where the agenda is always active should be special cases so that the first block is required to be the initial starting difficulty and then all subsequent blocks should use the first block as the anchor.

Reply:
I pushed a new commit (that will be merged after review) to implement the aforementioned behavior of special casing networks where the agenda is always active so that the first block is required to be the initial starting difficulty and then all subsequent blocks should use the first block as the anchor.

Reply:
The commit has been squashed.